Firestore/core/src/firebase/firestore/local/leveldb_mutation_queue.h

```diff
@@ -0,0 +1,152 @@
+/*
+ * Copyright 2019 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_LOCAL_LEVELDB_MUTATION_QUEUE_H_
+#define FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_LOCAL_LEVELDB_MUTATION_QUEUE_H_
+
+#if !defined(__OBJC__)
+#error "For now, this file must only be included by ObjC source files."
+#endif  // !defined(__OBJC__)
+
+#import <Foundation/Foundation.h>
+
+#include <set>
+#include <string>
+#include <vector>
+
+#import "Firestore/Source/Public/FIRTimestamp.h"
+
+#include "Firestore/core/src/firebase/firestore/auth/user.h"
+#include "Firestore/core/src/firebase/firestore/local/leveldb_key.h"
+#include "Firestore/core/src/firebase/firestore/model/document_key.h"
+#include "Firestore/core/src/firebase/firestore/model/document_key_set.h"
+#include "Firestore/core/src/firebase/firestore/model/types.h"
+#include "absl/strings/string_view.h"
+#include "leveldb/db.h"
+
+@class FSTLevelDB;
+@class FSTLocalSerializer;
+@class FSTMutation;
+@class FSTMutationBatch;
+@class FSTPBMutationQueue;
+@class FSTQuery;
+
+NS_ASSUME_NONNULL_BEGIN
+
+namespace firebase {
+namespace firestore {
+namespace local {
+
+/**
+ * Returns one larger than the largest batch ID that has been stored. If there
+ * are no mutations returns 0. Note that batch IDs are global.
+ */
+model::BatchId LoadNextBatchIdFromDb(leveldb::DB* db);
+
+class LevelDbMutationQueue {
+ public:
+  LevelDbMutationQueue(const auth::User& user,
+                       FSTLevelDB* db,
+                       FSTLocalSerializer* serializer);
+
+  void Start();
+
+  bool IsEmpty();
+
+  void AcknowledgeBatch(FSTMutationBatch* batch,
+                        NSData* _Nullable stream_token);
+
+  FSTMutationBatch* AddMutationBatch(FIRTimestamp* local_write_time,
+                                     NSArray<FSTMutation*>* mutations);
+
+  void RemoveMutationBatch(FSTMutationBatch* batch);
+
+  const std::vector<FSTMutationBatch*> AllMutationBatches();
```

Comment:
As above, since this accesses leveldb, our policy is not to mark it `const`.

Reply:
Ok. You should still consider dropping the first const, since it doesn't have any effect. i.e. this code works:
```
struct X { ... };
const X GetX() { ... };
void foo() {
  /* not const */ X x = GetX();
  x.MutableMethod();
}
```

If you feel the const still adds documentation value (along the lines of "I can't stop you, but please don't modify this value") then feel free to keep it, but if so, consider adding it to the other methods in this class too.

Reply:
Ah, got it. Removed.